pgmpy/factors/CPD.py

```diff
@@ -112,6 +112,27 @@ def __init__(self, variable, variable_card, values,
 
         super().__init__(variables, cardinality, values.flatten('C'))
 
+    def replace_variables(self, var1, var2):
+        new_scope = self.scope()
+        new_scope.reverse()
+        new_cardinality = self.cardinality[::-1].copy()
```

Comment:
@kislayabhi When you are just using ```new_cardinality = self.cardinality[::-1]``` you are basically doing a shallow copy. You can find more about about shallow copy and deep copy at https://docs.python.org/2/library/copy.html. So when you change the elements of ```new_cardinality``` then it would change the values of ```self.cardinality``` as ```new_cardinality``` contains the references of the objects present in ```self.cardinality```.

So I would suggest you to do something like this

```python
new_cardinality = np.array([var_card for var_card in self.cardinality[::-1]])
```

This way you are not creating a shallow copy rather you are creating a new array object with recursively adding the copy of the objects in ```self.cardinality``` so changing the elements of ```new_cardinality``` won't have any effect on ```self.cardinality```. 

Hope this makes sense :)